apps/els_lsp/src/els_bsp_provider.erl

```diff
@@ -0,0 +1,147 @@
+-module(els_bsp_provider).
+
+-behaviour(els_provider).
+
+%% API
+-export([ start/1 ]).
+
+%% els_provider functions
+-export([ is_enabled/0
+        , init/0
+        , handle_request/2
+        , handle_info/2
+        ]).
+
+
+%%==============================================================================
+%% Includes
+%%==============================================================================
+-include("els_lsp.hrl").
+-include_lib("kernel/include/logger.hrl").
+
+%%==============================================================================
+%% Types
+%%==============================================================================
+-type state() :: #{ running := boolean()          % is the BSP server running?
+                  , root_uri := uri() | undefined % the root uri
+                  , pending := list()             % pending requests
+                  }.
+
+%%==============================================================================
+%% API
+%%==============================================================================
+-spec start(uri()) -> ok | {error, term()}.
+start(RootUri) ->
+  els_provider:handle_request(?MODULE, {start, RootUri}).
+
+%%==============================================================================
+%% els_provider functions
+%%==============================================================================
+-spec init() -> state().
+init() ->
+  #{ running  => false
+   , root_uri => undefined
+   , pending  => []
+   }.
+
+-spec is_enabled() -> true.
+is_enabled() -> true.
+
+-spec handle_request({start, uri()}, state()) -> {ok, state()}.
+handle_request({start, RootUri}, #{ running := false } = State) ->
+  ?LOG_INFO("Starting BSP server in ~p", [RootUri]),
+  case els_bsp_client:start_server(RootUri) of
+    {ok, Config} ->
+      ?LOG_INFO("BSP server started from config ~p", [Config]),
+      enqueue(initialize_bsp),
```

Comment:
Do we need to do this asynchronously? It looks like the code handling the message makes an async request anyway?

Reply:
We could send the request right here, when I was writing it the intent was to separate starting (which is the current request) from initialization (which arguably is a separate step). I can combine the two here if you prefer that.